exploration/variable-mutability.md

```diff
@@ -0,0 +1,244 @@
+
+## Requirements
+
+_What properties does the solution have to manifest to enable the use-cases above?_
+
+- Be able to re-annotate variables without having to rename them in the message body
+- Allow static analysis to detect mistakes when referencing an undefined local variable
+- Be able to re-annotate variables multiple times (because we do not allow nesting)
+
+- _more needed?_
+
+## Constraints
+
+_What prior decisions and existing conditions limit the possible design?_
+
+- Variable names are potentially contrained by `Nmtoken`.
+  The reason we chose Nmtoken (and Name) was to maximize compatibility with (potential)
+  LDML constructs, since CLDR uses XML.
+  The "carve out" for various sigils doesn't conflict with naming in LDML currently and
+  future conflicts are under our (CLDR-TC's) control.
+
+## Proposed Design
+
+_Describe the proposed solution. Consider syntax, formatting, errors, registry, tooling, interchange._
+
+Separate local variables from externally passed values by altering the sigil
+and by using a visually distinctive pattern for local names
+(in an effort to prevent `$foo`/`@foo` confusion).
+
+```abnf
+variable     = local_var / external_var
+local_var    = "@_" name
+external_var = "$" name
+```
+
+> _Example_
+>
+> ```
+> let @_local = {$external :transform}
+> let @_anotherLocalVar = {|Some literal| :annotated}
+> ```
+
+To allow users to perform multiple annotations on a value,
+while still allowing detection of unintentional reassignment,
+introduce a new keyword `modify`:
+
+```abnf
+declaration = (let / modify) s variable [s] "=" [s] expression
+...
+modify = %6D.%6F.%64.%69.%66.%79
+```
+
+It is a syntax error to use `let` on a variable that has been previously
+assigned through any declaration (either `let` or `modify`)
+
+It is a variable resolution error to call `modify`
+on an external variable that does not exist.
+
+> _Example_
+>
+> ```
+> let @_local = {$external :transform}
+> modify @_local = {@_local :modification with=options}
+> modify $external = {$external :transform adding=annotation}
+> ```
+
+When more than one `modify` declaration applies to the same named variable,
+or when a `modify` declaration is applied to a local variable
+defined in a `let` declaration,
+the named variable behaves as if each declaration were called
+in the sequence in which they appear in the message.
+Implementations are not required (by this design, anyway)
+to resolve values in a greedy manner.
+They might not resolve a value unless it is actually used in a selector
+or in a placeholder.
+
+#### Sigil Choice for Local Variables
+
+The choice here of `@_` as the local variable sigil is probably not distinctive enough.
+It is probably okay to be a little inconvenient with local variable naming
+as these are less common than external variables.
+Alternatives to consider:
+
+- `@@foo`
+- `@foo@`
+- `@!foo`
+- `@ONLY_UPPER_ASCII_SNAKE`
```

Comment:
These are all pretty ugly, but I presume that's intentional. Do we not want users to use local variables? Because that's what we're saying by making them ugly and clumsy.

Reply:
Is there a beautiful way to make locale variables and not have them be confused with external variables "of the same name"? 

An alternative not presented is the "rope merchant" option: do nothing (ugly or otherwise) and let people adopt whatever convention prevents mistakes, e.g. `$foo` and `#foo` are separate values (external vs. local).

Note: I just recalled that we wanted to use `@` for annotation, so will change that to `#` globally.